vdev/src/features.rs

```diff
@@ -0,0 +1,128 @@
+use std::{collections::BTreeSet, collections::HashMap, ffi::OsStr, fs, path::Path};
+
+use anyhow::{bail, Context, Result};
+use serde::Deserialize;
+use serde_json::Value;
+
+type ComponentMap = HashMap<String, Component>;
+
+// Use a BTree to keep the results in sorted order
+type FeatureSet = BTreeSet<String>;
+
+/// This is a ersatz copy of the Vector config, containing just the elements we are interested in
+/// examining. Everything else is thrown away.
+#[derive(Deserialize)]
+pub struct VectorConfig {
+    api: Option<Value>,
+    enterprise: Option<Value>,
+
+    #[serde(default)]
+    sources: ComponentMap,
+    #[serde(default)]
+    transforms: ComponentMap,
+    #[serde(default)]
+    sinks: ComponentMap,
+}
+
+#[derive(Deserialize)]
+struct Component {
+    r#type: String,
+}
+
+macro_rules! mapping {
+    ( $( $key:ident => $value:ident, )* ) => {
+        HashMap::from([
+            $( (stringify!($key), stringify!($value)), )*
+        ])
+    };
+}
+
+pub fn load_and_extract(filename: impl AsRef<Path>) -> Result<String> {
+    let filename = filename.as_ref();
+    let config =
+        fs::read_to_string(filename).with_context(|| format!("failed to read {filename:?}"))?;
+
+    let config: VectorConfig = match filename.extension().and_then(OsStr::to_str) {
+        None => bail!("Invalid filename {filename:?}, no extension"),
+        Some("json") => serde_json::from_str(&config)?,
+        Some("toml") => toml::from_str(&config)?,
+        Some("yaml") => serde_yaml::from_str(&config)?,
```

Comment:
This seems like it would fail on `.yml` files?

Reply:
Will fix. Are there also shortened extensions I should check for the other two formats?

Reply:
I _think_ it's just the YAML one that has different extension options - could also cross-reference how Vector identifies them and make sure it's consistent.